Use package status instead of deleted bit

joelverhagen opened the pull request:
Address https://github.com/NuGet/NuGetGallery/issues/4645.

Depends on https://github.com/NuGet/NuGetGallery/pull/4664. I will change the base branch once that one goes in.

loic-sharma commented on src/NuGetGallery/Controllers/ODataV1FeedController.cs:
```diff
@@ -87,10 +88,10 @@ public IHttpActionResult GetCount(ODataQueryOptions<V1FeedPackage> options)
+                .Where(p => p.PackageRegistration.Id.Equals(id, StringComparison.OrdinalIgnoreCase) &&
+                            !p.IsPrerelease &&
+                            p.PackageStatusKey == PackageStatus.Available)
+                .Where(SemVerLevelKey.IsUnknown());
```
>IsUnknown [](start = 38, length = 9)

Nit: is there a better name to would indicate that `IsUnknown` returns *an expression* that checks if the package is unknown?

joelverhagen replied:
Renamed to `IsUnknownPredicate`.

loic-sharma commented on src/NuGetGallery/Services/PackageNamingConflictValidator.cs:
```diff
@@ -56,7 +56,7 @@ public bool IdConflictsWithExistingPackageTitle(string registrationId)
 
             return _packageRepository.GetAll()
-                .Any(p => !p.Deleted && p.Title.ToLower() == registrationId);
+                .Any(p => p.PackageStatusKey != PackageStatus.Deleted && p.Title.ToLower() == registrationId);
```
>Deleted [](start = 62, length = 7)

Consider adding a comment here that you MUST check more than just `Available` packages as this small nuance is crucial  :)

joelverhagen replied:
Sure

loic-sharma commented on src/NuGetGallery/Services/PackageService.cs:
```diff
@@ -470,7 +470,7 @@ where VersionRange.Parse(d.VersionSpec).Satisfies(packageVersion)
             }
 
-            if (package.Deleted)
+            if (package.PackageStatusKey == PackageStatus.Deleted)
```
>package.PackageStatusKey == PackageStatus.Deleted [](start = 16, length = 49)

Should this be `package.PackageStatusKey != PackageStatus.Active`? Or do we want users to list/unlist packages that vail validation?

joelverhagen replied:
List/unlist on "Validating" packages is supported, per @karann-msft's design.

loic-sharma commented on src/NuGetGallery/Services/PackageService.cs:
```diff
@@ -836,11 +836,11 @@ private void ValidatePackageTitle(PackageMetadata packageMetadata)
             var latestPackage = FindPackage(
                 packageRegistration.Packages,
-                p => !p.Deleted && p.Listed && p.SemVerLevelKey == SemVerLevelKey.Unknown);
+                p => p.PackageStatusKey == PackageStatus.Available && p.Listed && p.SemVerLevelKey == null);
```
>p.SemVerLevelKey == null [](start = 82, length = 24)

Shouldn't all these p.SemVerLevelKey be using `SemverLevelKey.IsUnknown`? If so, don't foget lines 843, 854, and 881

joelverhagen replied:
I matched existing `SemVerLevelKey` precedent. We cannot use `IsUnknown` inside of a `Func<Package, bool>`.

joelverhagen replied:
I'll see if it's worth refactoring.

joelverhagen replied:
Found a good solution.

skofman1 commented on src/NuGetGallery/Services/PackageNamingConflictValidator.cs:
```diff
@@ -55,8 +55,9 @@ public bool IdConflictsWithExistingPackageTitle(string registrationId)
+            // We allow any package status aside from soft-deleted to be considered for ID/title conflicts.
             return _packageRepository.GetAll()
-                .Any(p => !p.Deleted && p.Title.ToLower() == registrationId);
+                .Any(p => p.PackageStatusKey != PackageStatus.Deleted && p.Title.ToLower() == registrationId);
```
As discussed in the previous PR, we shouldn't assume the behavior here will be correct for future status keys, hence we should use an explicit subset

skofman1 replied:
We can leave it like this for now, and flip the condition if we ever introduce a status that's shouldn't block upload

shishirx34 commented on src/NuGetGallery.Core/SemVerLevelKey.cs:
```diff
@@ -8,18 +8,50 @@
     /// <summary>
-    /// Helper class to use to determine the SemVer level of a package version.
+    /// Helper class to use to determine the SemVer level of a package version. Hard-coded for now, but we can easily
+    /// expand to use an additional Sql table to join with when supporting additional semVerLevel's is needed.
```
nit: ...additional semverLevel as needed.

joelverhagen replied:
I think this comment is unnecessary actually. It's unlikely we will add a table for SemVer level.

joelverhagen replied:
Comment revised.

shishirx34 commented on src/NuGetGallery/Areas/Admin/Controllers/DeleteController.cs:
```diff
@@ -84,7 +84,7 @@ private DeleteSearchResult CreateDeleteSearchResult(Package package)
                 DownloadCount = package.DownloadCount,
                 Listed = package.Listed,
-                Deleted = package.Deleted
+                Deleted = package.PackageStatusKey == PackageStatus.Deleted,
```
So this code would be going in only after the migration from 'Deleted' => 'Package.StatusKey' is done?

joelverhagen replied:
Yes.

shishirx34 commented on src/NuGetGallery/Controllers/ODataV1FeedController.cs:
```diff
@@ -87,10 +88,10 @@ public IHttpActionResult GetCount(ODataQueryOptions<V1FeedPackage> options)
+                .Where(p => p.PackageRegistration.Id.Equals(id, StringComparison.OrdinalIgnoreCase) &&
+                            !p.IsPrerelease &&
+                            p.PackageStatusKey == PackageStatus.Available)
+                .Where(SemVerLevelKey.IsUnknownPredicate());
```
nit: the operators on subsequent lines rather than at the end, just for consistency.

joelverhagen replied:
I've seen both in similar frequency.

Regex | Hit count
--- | ---
` {4,}&&` | 91
` {4,}\|\|` | 30
`&&\s*$` | 99
`\|\|\s*$` | 21

joelverhagen replied:
Spoke offline. I like the logical operator at the beginning of the line better.

shishirx34 commented on src/NuGetGallery/Controllers/ODataV2FeedController.cs:
```diff
@@ -188,9 +188,9 @@ public class ODataV2FeedController
-                            && !p.Deleted)
-                .Where(SemVerLevelKey.IsPackageCompliantWithSemVerLevel(semVerLevel));
+                .Where(p => p.PackageStatusKey == PackageStatus.Available &&
+                            p.PackageRegistration.Id.Equals(id, StringComparison.OrdinalIgnoreCase))
```
nit: conditional operator on next line, maybe?

shishirx34 commented on src/NuGetGallery/Queries/AutoCompleteDatabasePackageVersionsQuery.cs:
```diff
@@ -14,7 +14,7 @@ public class AutoCompleteDatabasePackageVersionsQuery
 FROM Packages p (NOLOCK)
 	JOIN PackageRegistrations pr (NOLOCK) on pr.[Key] = p.PackageRegistrationKey
-WHERE p.[Deleted] <> 1 AND p.[Listed] = 1 AND {0} AND pr.ID = {{0}}
+WHERE p.[PackageStatusKey] = 0 AND p.[Listed] = 1 AND {0} AND pr.ID = {{0}}
```
may be use the `PackageStatus.Available` instead of the hard coded value here?

joelverhagen replied:
Good idea!

chenriksson commented:
The migration should correct existing data, to set PackageStatus.Deleted in cases of Package.Deleted. However, there's a window between when the migration ran and when the deployment finishes where packages could be deleted old-style. We should probably verify/correct data just after deployment as well, right?

joelverhagen commented:
> The migration should correct existing data, to set PackageStatus.Deleted in cases of Package.Deleted. However, there's a window between when the migration ran and when the deployment finishes where packages could be deleted old-style. We should probably verify/correct data just after deployment as well, right?

That's right. I plan that as part of the deployment steps. The alternative is to say "don't delete things" during that window. I prefer the former since it does not disrupt the support requests flow.

shishirx34 commented on src/NuGetGallery.Core/Auditing/AuditedEntities/AuditedPackage.cs:
```diff
@@ -86,6 +87,7 @@ public static AuditedPackage CreateFrom(Package package)
 #pragma warning disable CS0612 // Type or member is obsolete
                 Deleted = package.Deleted,
 #pragma warning restore CS0612 // Type or member is obsolete
+                HasReadMe = package.HasReadMe,
```
Why are these changes showing up here? Shouldn't this be in DEV already?

joelverhagen replied:
It is. I added a comma at the end of a line.
